### PR TITLE
Upload page

### PR DESCRIPTION
**Pull request by shad-k**
- Upload page UI
- Sign transaction
- Approve

Right now the request to store the meme metadata fails because of the new signature field added.
@sylar217 You'll need to check how to store the signature.

### CODE REVIEW

File: src/pages/Upload.tsx [hunk 1]
```diff
@@ -302,14 +265,137 @@ const Upload: React.FC<{}> = () => {
...
+
+            await contract.methods
+              .putOnSale(meme.tokenID)
+              .call({ from: authContext.authProvider?.account })
+              .then(() => {
+                console.log("Put on sale completed");
+              })
+              .catch(async function(error: any) {
+                console.log(error);
+                await textile.deleteMemeFromBucket(meme);
```
Comment (sylar217) [inline src/pages/Upload.tsx:335]: This is just putting the meme on sale right? Minting has already happened in another contract call, so we shouldn't delete the meme from textile.

File: src/pages/Upload.tsx [hunk 2]
```diff
@@ -302,14 +265,137 @@ const Upload: React.FC<{}> = () => {
...
+          {
+            method,
+            params,
+            from
+          },
+          async (err: any, result: any) => {
+            if (err) {
+              console.log(err);
+              const textile = await Textile.getInstance();
+              await textile.deleteMemeFromBucket(meme);
```
Comment (sylar217) [inline src/pages/Upload.tsx:312]: Even this is just signature for sale, minting has already happened, we shouldn't delete from backend.

File: src/pages/Upload.tsx [hunk 3]
```diff
@@ -302,14 +265,137 @@ const Upload: React.FC<{}> = () => {
...
+        contractAddress
+      );
+      return await contract.methods
+        //second paramenter is creator fee, using 0% for now
+        .setApprovalForAll(NetworkIDToSaleContract[networkId as number], true)
+        .send({ from: authContext.authProvider?.account })
+        .on("error", async function(error: any) {
+          console.log(error);
+          const textile = await Textile.getInstance();
+          await textile.deleteMemeFromBucket(meme);
```
Comment (sylar217) [inline src/pages/Upload.tsx:383]: Same, this is just part of sales, minting has already happened.

File: src/utils/Types.ts [hunk 4]
```diff
@@ -21,6 +23,7 @@ export interface MemeMetadata {
   dislikedBy: Array<string>; // set of user ids who disliked this meme.
   onSale?: boolean;
   description?: string;
+  sellApprovalSignature?: SignedTransaction;
```
[the remark this answers is not in the document]
Comment (matprime) [reply src/utils/Types.ts:26]: I talked to Lorenzo about that before, he said to me that signature is basically string. So I ques using string or hash string could work. Here are input parameters of buy smart contract function, which needs signature as input to check if money sent from buyer is correct. It need 3 parameters representing signature v,r,s:

    /**
     * @notice Buys token
     * @param tokenId       the id of the token to buy
     * @param price         the price of the token on sale
     * @param voters        array of voters on the meme
     * @param votes         array of votes of the voters
     * @param payCreator    defines if creator has to be paid or not
     * @param v             v part of the seller's signature
     * @param r             r part of the seller's signature
     * @param s             s part of the seller's signature
     */
    function buy(
        uint256 tokenId,
        uint256 price,
        address payable[] calldata voters,
        uint256[] calldata votes,
        bool payCreator,
        uint8 v,
        bytes32 r,
        bytes32 s
    )

**Review by matprime — CHANGES_REQUESTED** [review]
I have found 2 bugs after this commit:
1. When meme is created with price, browser developer console throws error when signing:
MetaMask - RPC Error: Provided chainId "3431" must match the active chainId "1337" 
There is also no Metamask window popping out to make a signature.
2. When is meme created and put on sale, dapp shows two copies of this meme

Update:
1. error was produced because of incorrect configuration in Metamask
2. Introducing fix for creation of duplicated memes with next commit

**Pull request merged**